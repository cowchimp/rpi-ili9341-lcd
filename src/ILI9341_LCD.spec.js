const { ILI9341_LCD } = require("./ILI9341_LCD");

describe("rgb888ToRgb565", () => {
  it("converts RGB array correctly", () => {
    const rpio = getMockRpio();
    const disp = new ILI9341_LCD(rpio, { width: 2, height: 4 });
    const singleBit = [0xff, 0x00, 0x00];
    const image = Buffer.alloc(
      disp.width * disp.height * singleBit.length,
      Buffer.from(singleBit)
    );
    const expected = Buffer.alloc(16, Buffer.from([0xf8, 0x00]));

    const result = disp.rgb888ToRgb565(image);

    expect(result).toEqual(expected);
  });

  it("converts RGBA array correctly", () => {
    const rpio = getMockRpio();
    const disp = new ILI9341_LCD(rpio, { width: 2, height: 4 });
    const singleBit = [0xff, 0x00, 0x00, 0xff];
    const image = Buffer.alloc(
      disp.width * disp.height * singleBit.length,
      Buffer.from(singleBit)
    );
    const expected = Buffer.alloc(16, Buffer.from([0xf8, 0x00]));

    const result = disp.rgb888ToRgb565(image, 4);

    expect(result).toEqual(expected);
  });
});

describe("showImage", () => {
  it("sets SPI data correctly", () => {
    const rpio = getMockRpio();
    const disp = new ILI9341_LCD(rpio, { width: 2, height: 4 });

    disp.Init();
    disp.showImage({
      width: 2,
      height: 4,
      data: Buffer.alloc(2 * 4 * 3, Buffer.from([0xff, 0x00, 0x00])),
    });
    disp.moduleExit();

    const data = rpio.spiWrite.mock.calls.map(([buff]) => buff.toString("hex"));
    expect(data).toMatchSnapshot();
  });

  it("throws if provided image dimensions does not match display dimensions", () => {
    const rpio = getMockRpio();
    const disp = new ILI9341_LCD(rpio, { width: 2, height: 4 });

    disp.Init();
    expect(() =>
      disp.showImage({
        width: 4,
        height: 2,
        data: Buffer.alloc(4 * 2 * 3, Buffer.from([0xff, 0x00, 0x00])),
      })
    ).toThrowError(
      "Image dimensions {4,2} do not match display dimensions {2,4}"
    );
  });
});

function getMockRpio() {
  const rpio = {
    init: jest.fn(),
    open: jest.fn(),
    spiBegin: jest.fn(),
    spiSetClockDivider: jest.fn(),
    spiEnd: jest.fn(),
    write: jest.fn(),
    sleep: jest.fn(),
    spiWrite: jest.fn(),
  };
  return rpio;
}
