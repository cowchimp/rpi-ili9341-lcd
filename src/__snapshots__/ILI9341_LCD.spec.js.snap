// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`showImage sets SPI data correctly 1`] = `
Array [
  "11",
  "cf",
  "00",
  "c1",
  "30",
  "ed",
  "64",
  "03",
  "12",
  "81",
  "e8",
  "85",
  "00",
  "79",
  "cb",
  "39",
  "2c",
  "00",
  "34",
  "02",
  "f7",
  "20",
  "ea",
  "00",
  "00",
  "c0",
  "1d",
  "c1",
  "12",
  "c5",
  "33",
  "3f",
  "c7",
  "92",
  "3a",
  "55",
  "36",
  "08",
  "b1",
  "00",
  "12",
  "b6",
  "0a",
  "a2",
  "44",
  "02",
  "f2",
  "00",
  "26",
  "01",
  "e0",
  "0f",
  "22",
  "1c",
  "1b",
  "08",
  "0f",
  "48",
  "b8",
  "34",
  "05",
  "0c",
  "09",
  "0f",
  "07",
  "00",
  "e1",
  "00",
  "23",
  "24",
  "07",
  "10",
  "07",
  "38",
  "47",
  "4b",
  "0a",
  "13",
  "06",
  "30",
  "38",
  "0f",
  "29",
  "36",
  "08",
  "2a",
  "00",
  "00",
  "00",
  "01",
  "2b",
  "00",
  "00",
  "00",
  "03",
  "2c",
  "f800f800f800f800f800f800f800f800",
]
`;
